jsonrpc2/codecs.go

```diff
@@ -26,31 +52,121 @@ var _ Codec = &jsonCodec{}
 // IOCodec returns a Codec that wraps JSON encoding and decoding over IO.
 func IOCodec(rwc io.ReadWriteCloser) *jsonCodec {
 	return &jsonCodec{
-		rwc: rwc,
+		encoder: json.NewEncoder(rwc),
+		decoder: json.NewDecoder(rwc),
+		closer:  rwc,
 	}
 }
 
 type jsonCodec struct {
-	rwc        io.ReadWriteCloser
+	encoder *json.Encoder
+	decoder *json.Decoder
+	closer  io.Closer
+
 	remoteAddr string
+
+	mu          sync.Mutex
+	batchBuffer []Message
 }
 
 func (codec *jsonCodec) RemoteAddr() string {
 	return codec.remoteAddr
 }
 
-func (codec *jsonCodec) ReadMessage() (*Message, error) {
-	var msg Message
-	err := json.NewDecoder(codec.rwc).Decode(&msg)
-	return &msg, err
+func (codec *jsonCodec) WriteMessage(msg *Message) error {
+	return codec.encoder.Encode(msg)
 }
 
-func (codec *jsonCodec) WriteMessage(msg *Message) error {
-	return json.NewEncoder(codec.rwc).Encode(msg)
+func (codec *jsonCodec) WriteBatch(batch []Message) error {
+	return codec.encoder.Encode(batch)
+}
+
+// ReadMessage supports consuming batched messages. When ReadMessage is
+// called, batched messages are automatically converted into single message
+// streams, but the associated Replier takes care of writing it back into the
+// batch.
+func (codec *jsonCodec) ReadMessage() (*Message, error) {
+	// Do we already have a message in the buffer?
+	codec.mu.Lock()
+	if len(codec.batchBuffer) > 0 {
+		// Pop
+		var msg Message
+		msg, codec.batchBuffer = codec.batchBuffer[len(codec.batchBuffer)-1], codec.batchBuffer[:len(codec.batchBuffer)-1]
+		codec.mu.Unlock()
+		return &msg, nil
+	}
+	codec.mu.Unlock()
+
+	// Get new messages. We load the next sequence of messages into a
+	// json.RawMessage so we can differentiate between a batched and unbatched
+	// message without borking the json.Decoder internal state.
+
+	// FIXME: Consider using a different streaming parser so we could do it all
+	// in one pass without pre-loading everything into memory and doing two
+	// passes.
+	var raw json.RawMessage
+	if err := codec.decoder.Decode(&raw); err != nil {
+		return nil, err
+	}
+
+	if !isArray(raw) {
+		// Single message, skip the batch parsing
+		var msg Message
+		if err := json.Unmarshal(raw, &msg); err != nil {
+			return nil, err
+		}
+
+		if msg.Request != nil && msg.ID != nil {
+			// Associate a replier on requests that are not notifications.
+			msg.Request.replier = &reply{
+				codec: codec,
+				id:    msg.ID,
+			}
+		}
+		return &msg, nil
+	}
+
+	// Batch message parsing time!
+	var batch []Message
+	if err := json.Unmarshal(raw, &batch); err != nil {
+		return nil, err
+	}
+
+	if len(batch) == 0 {
+		// FIXME: Should we be forgiving of this? Just return an empty message,
+		// or keep blocking until we get one?
+		return nil, errors.New("empty message batch")
```

Comment:
According to the spec:

> If the batch rpc call itself fails to be recognized as an valid JSON or as an Array with at least one value, the response from the Server MUST be a single Response object

https://www.jsonrpc.org/specification#batch

Reply:
Also from the examples:

```
--> []
<-- {"jsonrpc": "2.0", "error": {"code": -32600, "message": "Invalid Request"}, "id": null}
```

Reply:
Good catch on that in the specs, I missed it!

I'm preeeetty sure the outcome will be correct, because the error at the codec level will get converted into a JSONRPC error message at the transport level.

I should totally add a test for this.

Interestingly, it seems geth's jsonrpc implementation ignores empty batches.

Reply:
In fact, I should add all the examples as tests.

Reply:
You were right, needed to wire up the actual response error. Should be fixed in bca7459